rs/pocket_ic_server/src/pocket_ic.rs

```diff
@@ -221,10 +228,160 @@ impl Drop for BitcoinAdapterParts {
+    }
+}
+
+pub(crate) type CanisterHttpClient = Arc<
+    Mutex<
+        Box<
+            dyn NonBlockingChannel<
+                    AdapterCanisterHttpRequest,
+                    Response = AdapterCanisterHttpResponse,
+                > + Send,
+        >,
+    >,
+>;
+
+pub(crate) struct CanisterHttp {
+    pub client: CanisterHttpClient,
+    pub pending: Vec<CanisterHttpRequestId>,
+}
+
+pub(crate) struct Subnet {
+    pub state_machine: Arc<StateMachine>,
+    pub canister_http: Arc<Mutex<CanisterHttp>>,
+    delegation_from_nns: Arc<OnceCell<CertificateDelegation>>,
+    _canister_http_adapter_parts: CanisterHttpAdapterParts,
+}
+
+impl Subnet {
+    fn new(state_machine: Arc<StateMachine>) -> Self {
+        let uds_path = NamedTempFile::new().unwrap().into_temp_path().to_path_buf();
+        let log_level = state_machine.log_level;
+        let canister_http_adapter_parts = CanisterHttpAdapterParts::new(
+            uds_path.clone(),
+            log_level,
+            state_machine.replica_logger.clone(),
+            state_machine.metrics_registry.clone(),
+            state_machine.runtime.clone(),
+        );
+        let adapter_config = AdaptersConfig {
+            https_outcalls_uds_path: Some(uds_path),
+            ..Default::default()
+        };
+        let delegation_from_nns = Arc::new(OnceCell::new());
+        let client = setup_canister_http_client(
+            state_machine.runtime.handle().clone(),
+            &state_machine.metrics_registry,
+            adapter_config,
+            state_machine.query_handler.lock().unwrap().clone(),
+            MAX_CANISTER_HTTP_REQUESTS_IN_FLIGHT,
+            state_machine.replica_logger.clone(),
+            state_machine.get_subnet_type(),
+            delegation_from_nns.clone(),
+        );
+        let canister_http = Arc::new(Mutex::new(CanisterHttp {
+            client: Arc::new(Mutex::new(client)),
+            pending: vec![],
+        }));
+        Self {
+            state_machine,
+            canister_http,
+            delegation_from_nns,
+            _canister_http_adapter_parts: canister_http_adapter_parts,
+        }
+    }
+
+    fn set_delegation_from_nns(&self, delegation_from_nns: CertificateDelegation) {
+        self.delegation_from_nns.set(delegation_from_nns).unwrap();
+    }
+}
+
+pub(crate) struct SubnetsImpl {
+    subnets: Arc<RwLock<BTreeMap<SubnetId, Arc<Subnet>>>>,
+}
+
+impl SubnetsImpl {
+    fn new() -> Self {
+        Self {
+            subnets: Arc::new(RwLock::new(BTreeMap::new())),
+        }
+    }
+    fn get_subnet(&self, subnet_id: SubnetId) -> Option<Arc<Subnet>> {
+        self.subnets.read().unwrap().get(&subnet_id).cloned()
+    }
+    pub(crate) fn get_all(&self) -> Vec<Arc<Subnet>> {
+        self.subnets.read().unwrap().values().cloned().collect()
+    }
+}
+
+impl Subnets for SubnetsImpl {
+    fn insert(&self, state_machine: Arc<StateMachine>) {
+        self.subnets.write().unwrap().insert(
```

Comment:
Is it safe to unwrap here?

Reply:
We're already unwrapping when acquiring locks everywhere: it is only unsafe if some thread holding the lock panicked and this lock got poisoned due to that.

Reply:
I see!